Gutenberg Block for Tweet Shortcode?

georgestephanis opened the pull request:
Work In Progress!

georgestephanis committed:
A good start maybe?

aduth commented on modules/shortcodes/tweet.php:
```diff
@@ -142,4 +143,61 @@ static public function jetpack_tweet_shortcode_script() {
+		category: 'layout',
+
+		attributes : {
+			tweet : wp.blocks.query.query( 'input[name=tweet]').value
```
The idea with [attribute matchers](http://gutenberg-devdoc.surge.sh/reference/attribute-matchers/) (or, perhaps soon to be called ["sourced" values](https://github.com/WordPress/gutenberg/pull/1902)) is to avoid duplicating data between [comment attributes](https://github.com/WordPress/gutenberg/blob/master/docs/faq.md#how-is-data-stored-ive-seen-html-comments-what-is-their-purpose) and the actual saved markup of the block. For example, we don't need to serialize an image's URL into the comments blob if we know we can easily grab it from the `img`'s `src` attribute. In more complex examples, duplication is inevitable or there's a need for hidden metadata.

So what we define in `attributes` purely relates to _getting an object representation of the block_ back from the saved content, nothing to do with the `edit`-able render.

In your case, the saved content is a shortcode, from which you want to extract the `tweet` value. There's not a matcher for this (arguments could be made either way for whether we should [add conveniences for shortcodes](https://github.com/WordPress/gutenberg/issues/334)), so you might just want to let this live in comment metadata. Currently this is automatic if you set an attribute from your `edit` form but don't define it in the `attributes` property, but this is [likely to change soon](https://github.com/WordPress/gutenberg/pull/1905) to be more explicit (i.e. define all attributes, both the ones sourced from content and those serialized into comments).

I'm supposing you were expecting that defining this here would establish a link between the attribute value and the input from `edit`, which is not the case. Instead, you'll want to explicitly call `props.setAttributes` from your `edit` function when the input changes, similar to the concept of a [controlled input](https://facebook.github.io/react/docs/forms.html#controlled-components) in React.

```js
onInput: function( event ) {
	props.setAttributes({
		tweet: event.target.value
	});
}
```

georgestephanis replied:
Thanks for the explanation.  I wasn't expecting anything really, just playing around and seeing how things connect.

Regarding extracting stuff out -- can the attribute matcher / sourced values be a custom function passed in that will return what we care about from the contents somehow?  If so, I could just call `wp.shortcode.next()` on the contents of the block to extract and parse the shortcode and its parameters -- I've already enqueued `shortcodes.js` on the page for later usage building it.

aduth replied:
Currently this is possible by defining `attributes` as a function, which is passed the raw content (string) of the block and expected to return an object of attributes.

https://github.com/WordPress/gutenberg/blob/33542dd/blocks/api/parser.js#L28-L29
https://github.com/WordPress/gutenberg/blob/master/blocks/README.md#wpblocksregisterblocktype-name-string-typedefinition-object-

However, with increasing pressure to introduce server-capable attributes schemas / parsing, I'm not sure whether this can be relied upon.

Some more discussion at: https://github.com/WordPress/gutenberg/pull/1905

aduth commented on modules/shortcodes/tweet.php:
```diff
@@ -142,4 +143,61 @@ static public function jetpack_tweet_shortcode_script() {
+				{
+					name : 'tweet',
+					type : 'url',
+					value : props.tweet
```
`props` isn't the attributes of the block, but attributes can be accessed from `props.attributes`. There are a number of other values in props as well, such as `setAttributes` noted in the previous comment.

Instead, this would be: `props.attributes.tweet`

aduth commented on modules/shortcodes/tweet.php:
```diff
@@ -142,4 +143,61 @@ static public function jetpack_tweet_shortcode_script() {
+				attrs   : {
+					named   : {},
+					numeric : [
+						props.tweet
```
As above, this would be: `props.attributes.tweet`

aduth commented on modules/shortcodes/tweet.php:
```diff
@@ -142,4 +143,61 @@ static public function jetpack_tweet_shortcode_script() {
+		};
+
+	wp.blocks.registerBlockType( 'jetpack/tweet', {
+		title: wp.i18n.__( 'Tweet', 'jetpack' ),
```
A text domain solution hasn't yet been implemented for `__`, so the second parameter has no effect here.

georgestephanis replied:
Got it.  I was just basing this off prior art in https://github.com/nylen/gutenberg-examples/blob/master/01-basic/block.js#L20

aduth replied:
Created an issue at https://github.com/nylen/gutenberg-examples/issues/9

georgestephanis committed:
Correct how we refer to the tweet attribute.

georgestephanis commented:
Also, I'm not sure why, but my `wp:jetpack/tweet` blocks keep changing to `wp:core/freeform` on reloading the editor.

georgestephanis commented:
I think ^^ may be happening because I'm registering my block too late perhaps, so it's getting reset to core/freeform?

georgestephanis committed:
Update the prop when the field changes.  Still not thrilled with this method.

georgestephanis committed:
Add a dummy function to explore how we can parse this on initialization.

georgestephanis committed:
Somewhat tidier way of doing the script.

This lets it be inline on the page, with the other scripts -- rather
than printed at the very bottom.

Leaving the `script` tags there but commented out so PHPStorm still
syntax highlights it as js.

georgestephanis commented:
https://github.com/Automattic/jetpack/pull/7597/commits/6817940b28fd178c106303f1439fc4d4930829ec resolves the timing issue -- we're not printing it out manually any more, so it just goes at the right time via the core script system.

georgestephanis commented:
Somewhat interestingly, to get the dummy script to be enqueued, it has to be registered first and then seperately enqueued -- it can't just be enqueued as one step when the `$src` is falsey.

aduth commented:
It might be good to sort through issues of adding a block registration inline, but is there a particular reason you've chosen not to create a separate `.js` file for your block and enqueue that?

georgestephanis committed:
Clear out some unused chaff.

georgestephanis committed:
Flesh out how to find the tweet url from the node.

georgestephanis committed:
Don't return the result of the array directly.

Build, then compile.  This will make it easier for us to add more
conditional additions later.

georgestephanis committed:
Use onChange not onInput to satisfy browser console nag.

georgestephanis committed:
Add in handling for other optional attributes.

georgestephanis committed:
Wrap it in the generated classname?  Probably a good idea?

georgestephanis committed:
Oops, forgot the closing parens.

georgestephanis committed:
Set the generated shortcode to single.

georgestephanis committed:
Change to an array so we can return multiples.

georgestephanis committed:
Add BlockControls AlignmentToolbar.

georgestephanis committed:
Add sidebar width control.

georgestephanis committed:
If there's a className to be set, wrap it in a div so it can be set on that.

georgestephanis committed:
This is necessary so React doesn't turn `"` into `&quot;`

georgestephanis committed:
Rewrite attributes array to a function.

georgestephanis committed:
Are you the keymaster?

https://www.youtube.com/watch?v=pNnIbFFzpD8

georgestephanis commented:
I'm also getting this error:

> Warning: A component is changing an uncontrolled input of type url to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components

georgestephanis committed:
Alias `el` to `wp.element.createElement`

For brevity.

aduth commented:
This is because, since your attributes aren't being parsed, the default value for what you're assigning into the input is `undefined`, but then as the user types into the input is assigned a value. React treats an `undefined` value as an uncontrolled input and an non-empty value as a controlled input, and warns when you switch between the two. The workaround is to ensure that you never assign `undefined` as the value: either by making sure the attribute has a default, or providing a fallback value in the `value: attributes.foo || ''`

dereksmart commented:
Closing, as stale and will be picked up by another team or by George when he returns
